### ISSUE TITLE
Extend Studio usage reporting to support referenced fields in addition to executed fields

### ISSUE THREAD

**Issue by hwillson** [apollographql/apollo-server#5708]
In Apollo Studio, the "Fields" page lets you see how often your fields are *executed* by operations — ie, how often their resolvers run. It additionally lets you see which clients and operations ran those operations.

However, knowing if an operation *executed* a field doesn't tell the whole story about the relationship between operations and fields.  You might also be curious to learn if a field was textually referenced in the operation itself.

It's possible for a field to be referenced without executing. In fact, there are many reasons this can happen:
- The field is nested under another field which evaluates to null
- The field is nested under another field which evaluates to an empty list
- The field is nested under a non-matching fragment
- The field is nested under `@include` or `@skip`

If you're using the Fields page to determine "what are all the operations that use this field", you probably want to know about these usages too!

It's also possible for a field to be executed without being referenced. That's because when we track "executed fields", we are always tracking the concrete object type that is being executed, not an interface type it may have been resolved through.  So with schema:

```graphql
interface Animal {
  legs: Int
}
type Dog implements Animal {
  legs: Int
}
type Query {
  myFavoriteAnimal: Animal  # In practice, always returns a Dog
}
```

the operation `{ myFavoriteAnimal { legs } }` "references" `Animal.legs`, not `Dog.legs`, but it is the field `Dog.legs` that is executed.

We aim to improve Apollo Studio and the apollo-server usage reporting so that the Fields page displays both the number of executions *and* the number of textual references.

In addition to this just being an interesting thing to ask about, this will also mean that the Fields page can be useful without needing full execution tracing. There is a real performance impact of full execution tracing, especially in the federation context where the ftv1 protocol involves sending subgraph traces to the gateway/router in-band in the response.  And not every GraphQL server supports full execution tracing in the first place.  If we support referenced fields on the Fields page, then you can run a gateway/router in front of an arbitrary subgraph without federated tracing (whether you want to do this for performance or lack of implementation reasons) and still get some useful data on the Fields page.  (You can then perhaps run full tracing on a sampled subset of queries to get reasonable approximate execution data too.)

**Comment by glasser** [comment]
Just as an external status update: this project had two halves. One is an internal refactoring to improve how related data is stored in our backend databases, and that has now been implemented. The next step is to make the Apollo Server related changes to report field usage differently. This won't happen until October.

**Comment by glasser** [comment]
@RichardWright That sounds like something you could build with the plugin API? Seems pretty separate from Studio usage reporting.

**Comment by alexandernanberg** [comment]
Will you be able to use this without Apollo Studio? E.g. if you would like to use a custom server solution?

**Comment by glasser** [comment]
This is an improvement to a feature of Apollo Studio. You certainly could use the open source usage reporting plugin as a basis for reporting to something else if you'd like, but that's not a goal of this project.

**Comment by glasser** [comment]
This is released in Apollo Server 3.6.0.
